spec/models/weapon_spec.rb

```diff
@@ -0,0 +1,606 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Weapon, type: :model do
+  describe 'validations' do
+    subject(:validate) { weapon.validate }
+
+    let(:weapon) { build(:weapon) }
+
+    it 'is invalid without a name' do
+      weapon.name = nil
+      validate
+      expect(weapon.errors[:name]).to include "can't be blank"
+    end
+
+    it 'is invalid with an invalid category name' do
+      weapon.category = 'foo'
+      validate
+      expect(weapon.errors[:category]).to include 'must be "one-handed", "two-handed", or "archery"'
+    end
+
+    it 'is invalid with an invalid weapon type' do
+      weapon.weapon_type = 'foo'
+      validate
+      expect(weapon.errors[:weapon_type]).to include 'must be a valid type of weapon that occurs in Skyrim'
+    end
+
+    it 'is invalid with a negative unit weight' do
+      weapon.unit_weight = -0.5
+      validate
+      expect(weapon.errors[:unit_weight]).to include 'must be greater than or equal to 0'
+    end
+  end
+
+  describe '::before_validation' do
+    subject(:validate) { weapon.validate }
+
+    context 'when there is a canonical model assigned' do
+      let(:canonical_weapon) { weapon.canonical_weapon }
+
+      context 'when the canonical model has no enchantments' do
```

Comment:
This context is not really required since the next context tests the same things.